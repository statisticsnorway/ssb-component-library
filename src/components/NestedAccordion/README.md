NestedAccordion
========

> Nested accordion

### Usage

#### HTML

```html
<div class="ssb-nested-accordion mt-3">
    <button class="nested-accordion-header closed" onclick="{toggle classname to 'open'}">
        <span class="button-grid">
			<!-- If open -->
			<i class="expand-icon">{feather.chevronUp 16px}</i>
			<!-- If closed -->
			<i class="expand-icon">{feather.chevronDown 16px}</i>
			<span class="header-text">A nested accordion header</span>
        </span>
    </button>
    <div class="accordion-body closed">
		{insert content}
	</div>
</div>
```
Toggle the all instances of the `closed` className to `open` when clicked.

#### React

```jsx harmony
<Accordion primary header="This is a primary header">
  This is paragraph text which explains the accordion, the rest of the text is just to fill it out and show the space it takes.
  <NestedAccordion header="A nested accordion header">
    This is paragraph text which explains the accordion, the rest of the text is just to fill it out and show the space it takes.
  </NestedAccordion>
</Accordion>
```

Available props:

| Name       | Type           | Description  |
| ---------- | ------------- | ----- |
| children | node | Accordion content |
| header | string | Header text |
| openByDefault | bool | Will set the open state on init |

