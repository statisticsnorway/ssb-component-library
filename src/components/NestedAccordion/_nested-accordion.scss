.ssb-nested-accordion {
  display: grid;
  min-width: 300px;
  width: 100%;

  .nested-accordion-header {
    @include reset-button;

    .button-grid {
      align-items: center;
      cursor: pointer;
      display: grid;
      grid-column-gap: 10px;
      grid-template-columns: 10px 1fr;
    }

    .expand-icon {
      color: $ssb-green-3
    }

    &:hover {

      .header-text {
        color: $ssb-green-4;
      }
    }

    &.open {

      .header-text, .sub-header {
        color: $ssb-green-4;
      }

      .expand-icon {
        color: $ssb-green-4;
      }
    }
  }

  .nested-accordion-body {
    font-size: 16px;
    padding: 10px 10px 10px 20px;

    &.closed {
      display: none;
      height: 0;
    }
  }
}
