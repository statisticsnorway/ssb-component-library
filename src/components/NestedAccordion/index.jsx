import React, { useState } from 'react';
import PropTypes from 'prop-types';
import { Plus, Minus } from 'react-feather';

const NestedAccordion = ({
	children, header, openByDefault,
}) => {
	const [isOpen, toggleOpen] = useState(openByDefault);
	return (
		<div className={`ssb-nested-accordion mt-3`}>
			<div
				className={`nested-accordion-header ${isOpen ? 'open' : 'closed'}`}
				onClick={() => toggleOpen(!isOpen)}
			>
				{!isOpen && <Plus className="expand-icon" size="10" />}
				{isOpen && <Minus className="expand-icon" size="10" />}
				<span className="header-text">{header}</span>
			</div>
			<div className={`nested-accordion-body ${isOpen ? 'open' : 'closed'}`}>
				{isOpen && children}
			</div>
		</div>
	);
};

NestedAccordion.defaultProps = {
	openByDefault: false,
};

NestedAccordion.propTypes = {
	children: PropTypes.node,
	header: PropTypes.string,
	openByDefault: PropTypes.bool,
};

export default NestedAccordion;
