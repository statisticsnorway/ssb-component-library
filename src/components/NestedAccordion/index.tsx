import React, { useEffect, useRef, useState, type ReactNode } from 'react'
import { ChevronDown, ChevronUp } from 'react-feather'
import classNames from '../../utils/utils'

interface NestedAccordionProps {
  children?: ReactNode
  className?: string
  header?: string
  openByDefault?: boolean
  onToggle?: (isOpen: boolean) => void
}

const NestedAccordion: React.FC<NestedAccordionProps> = ({
  children,
  className,
  header,
  openByDefault = false,
  onToggle = () => {},
}) => {
  const [isOpen, setIsOpen] = useState(openByDefault)
  const firstUpdate = useRef(true)
  useEffect(() => {
    if (firstUpdate.current) {
      firstUpdate.current = false
      return
    }
    onToggle(isOpen)
  }, [isOpen])
  return (
    <div className={classNames('ssb-nested-accordion', className)}>
      <button
        className={`nested-accordion-header ${isOpen ? 'open' : 'closed'}`}
        aria-expanded={isOpen ? 'true' : 'false'}
        onClick={() => setIsOpen(!isOpen)}
      >
        <span className='button-grid'>
          {!isOpen && <ChevronDown className='expand-icon' size={16} />}
          {isOpen && <ChevronUp className='expand-icon' size={16} />}
          <span className='header-text'>{header}</span>
        </span>
      </button>
      <div className={`nested-accordion-body ${isOpen ? 'open' : 'closed'}`}>{children}</div>
    </div>
  )
}

export default NestedAccordion
