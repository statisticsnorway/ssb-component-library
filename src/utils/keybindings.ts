export const KEY_ARROW_LEFT = 37
export const KEY_ARROW_UP = 38
export const KEY_ARROW_RIGHT = 39
export const KEY_ARROW_DOWN = 40

export const KEY_ESCAPE = 27
export const KEY_ENTER = 13
export const KEY_BACKSPACE = 8
export const KEY_TAB = 9
export const KEY_SPACE = 32

export const KEY_SHIFT = 16
export const KEY_CTRL = 17
export const KEY_ALT = 18
export const KEY_META = 91
