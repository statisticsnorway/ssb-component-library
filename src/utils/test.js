import fs from 'fs';
import path from 'path';
import { render as rtlRender } from '@testing-library/react';

export * from '@testing-library/react';
export * from '@testing-library/jest-dom';

const styles = fs.readFileSync(path.relative(process.cwd(), './lib/bundle.css'), 'utf8');

export function render(ui, options) {
	const view = rtlRender(ui, options);

	const styleElement = document.createElement('style');
	styleElement.innerHTML = styles;
	document.body.appendChild(styleElement);

	return view;
}
