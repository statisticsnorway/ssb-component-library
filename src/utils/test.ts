import fs from 'fs'
import path from 'path'
import { RenderOptions, RenderResult, render as rtlRender } from '@testing-library/react'
import { ReactNode } from 'react'

export * from '@testing-library/react'

const styles = fs.readFileSync(path.relative(process.cwd(), './lib/bundle.css'), 'utf8')

export function render(ui: ReactNode, options?: RenderOptions): RenderResult {
  const view = rtlRender(ui, options)

  const styleElement = document.createElement('style')
  styleElement.innerHTML = styles
  document.body.appendChild(styleElement)

  return view
}
