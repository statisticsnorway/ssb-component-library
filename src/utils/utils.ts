// utils.ts
const classNames = (...classes: (string | boolean | undefined)[]): string => classes.filter(Boolean).join(' ')

export default classNames
