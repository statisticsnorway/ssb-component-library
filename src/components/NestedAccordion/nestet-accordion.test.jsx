import React from 'react'
import userEvent from '@testing-library/user-event'

import { screen, render } from '../../utils/test'
import NestedAccordion from './index'

describe('NestedAccordion component', () => {
  test('Matches the snapshot', () => {
    const { asFragment } = render(
      <NestedAccordion header='A nested accordion header'>Nestet accordion</NestedAccordion>
    )
    expect(asFragment()).toMatchSnapshot()
  })
  test('Change to open on click', async () => {
    const user = userEvent.setup()
    const headerText = 'A nested accordion header'
    const bodyText = 'Nestet accordion'
    render(<NestedAccordion header={headerText}>{bodyText}</NestedAccordion>)
    const wrapper = screen.getByText(headerText)

    await user.click(wrapper.parentElement)

    expect(await screen.findByText(bodyText)).toBeVisible()
  })
})
