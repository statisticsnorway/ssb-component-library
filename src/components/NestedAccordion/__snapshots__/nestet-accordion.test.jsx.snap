// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NestedAccordion component Matches the snapshot 1`] = `
<DocumentFragment>
  <div
    class="ssb-nested-accordion"
  >
    <button
      aria-expanded="false"
      class="nested-accordion-header closed"
    >
      <span
        class="button-grid"
      >
        <svg
          class="expand-icon"
          fill="none"
          height="16"
          stroke="currentColor"
          stroke-linecap="round"
          stroke-linejoin="round"
          stroke-width="2"
          viewBox="0 0 24 24"
          width="16"
          xmlns="http://www.w3.org/2000/svg"
        >
          <polyline
            points="6 9 12 15 18 9"
          />
        </svg>
        <span
          class="header-text"
        >
          A nested accordion header
        </span>
      </span>
    </button>
    <div
      class="nested-accordion-body closed"
    >
      Nestet accordion
    </div>
  </div>
</DocumentFragment>
`;
